### PR TITLE
Bakerstu mcp2515 improvements

### PR DESCRIPTION
**Pull request by bakerstu**
Ready for final peer review.

### CODE REVIEW

File: src/freertos_drivers/common/MCP2515Can.hxx [hunk 1]
```diff
@@ -330,137 +344,182 @@ private:
...
-            uint8_t unused3 : 2; /**< unused bits */
-            uint8_t rtr     : 1; /**< remote transmit request bit */
-            uint8_t unused4 : 1; /**< unused bit */
+            uint8_t dlc_     : 4; /**< data length code */
+            uint8_t unused3_ : 2; /**< unused bits */
+            uint8_t rtr_     : 1; /**< remote transmit request bit */
+            uint8_t unused4_ : 1; /**< unused bit */
         };
-        uint8_t data[8]; /** all 8 data bytes */
+        uint8_t data_[8];          /** all 8 data bytes */
```
Comment (bakerstu) [inline src/freertos_drivers/common/MCP2515Can.hxx]: Add a static_assert() in order to validate the alignment of data_

File: src/freertos_drivers/ti/CC32xxSPI.cxx [hunk 2]
```diff
@@ -136,123 +132,138 @@ int CC32xxSPI::update_configuration()
...
-/// Debugging helper.
-volatile uint32_t g_status;
-
-/** Common interrupt handler for all SPI devices.
+/** Configure a DMA transaction.
+ * @param msg message to transact.
  */
-void CC32xxSPI::interrupt_handler()
+__attribute__((optimize("-O3")))
+void CC32xxSPI::config_dma(struct spi_ioc_transfer *msg)
```
Comment (bakerstu) [inline src/freertos_drivers/ti/CC32xxSPI.cxx]: Need to check build under legacy CC3200 target.

File: src/freertos_drivers/common/SPI.hxx [hunk 3]
```diff
@@ -54,6 +55,49 @@ public:
...
     typedef void (*ChipSelectMethod)();
 
+    /** Method to transmit/receive the data.  Chip select will be asserted at
+     * start of the transfer and deasserted at the end of the transfer.
+     * @param self pointer to a SPI object instance
+     * @param msg message(s) to transact.
+     * @return bytes transfered upon success, -errno upon failure
+     */
+    __attribute__((optimize("-O3")))
+    static int transfer(SPI *self,struct spi_ioc_transfer *msg)
```
Comment (balazsracz) [inline src/freertos_drivers/common/SPI.hxx]: let's not make this static but a regular instance function.
if there is a difference in assertion behavior, let's include that int he function name, like transfer_with_cs_assert.

File: include/compiler.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * @author Stuart W. Baker
+ * @date 10 June 2018
+ */
+
+#ifndef _INCLUDE_COMPILER_H_
+#define _INCLUDE_COMPILER_H_
+
+#if defined (__GNUC__)
+    #define LIKELY(x)   __builtin_expect(!!(x), 1)
```
Comment (balazsracz) [inline include/compiler.h]: please check that !! does not turn into any assembly instructions.

File: src/freertos_drivers/common/SPI.hxx (hunk 3 above)
Comment (bakerstu) [reply src/freertos_drivers/common/SPI.hxx]: Fixed.

File: src/freertos_drivers/common/MCP2515Can.hxx (hunk 1 above)
Comment (bakerstu) [reply src/freertos_drivers/common/MCP2515Can.hxx]: Fixed.

File: src/freertos_drivers/common/MCP2515Can.hxx [hunk 5]
```diff
@@ -330,137 +345,184 @@ private:
...
+                sidh_ = (can_frame->can_id & 0x1FE00000) >> 21;
+                eid_ = (can_frame->can_id & 0x00030000) >> 16;
+                exide_ = 1;
+                sid_ = (can_frame->can_id & 0x001C0000) >> 18;
+                eid8_ = (can_frame->can_id & 0x0000FF00) >> 8;
+                eid0_ = (can_frame->can_id & 0x000000FF) >> 0;
+            }
+            else
+            {
+                sidh_ = (can_frame->can_id & 0x000007F8) >> 3;
```
Comment (balazsracz) [inline src/freertos_drivers/common/MCP2515Can.hxx]: there is a bug here. This line cannot be the same as line 448.

File: src/freertos_drivers/common/MCP2515Can.hxx [hunk 6]
```diff
@@ -330,137 +345,184 @@ private:
...
+    /** Setup a buffer read transfer structure.
+     */
+    class BufferRead : public Buffer
+    {
+    public:
+        /** Constructor.
+         * @param index buffer index to read from (valid values are 0 and 1)
+         */
+        BufferRead(int index)
+            : Buffer(READ_RX_BUF | (index == 0 ? 0x00 : 0x04))
```
Comment (balazsracz) [inline src/freertos_drivers/common/MCP2515Can.hxx]: do you have a parallel to the LOAD_TX_BUF constant that you could use here?

File: src/freertos_drivers/common/SPI.hxx [hunk 7]
```diff
@@ -54,6 +55,78 @@ public:
...
+        {
+            count += msgs->len;
+            csAssert();
+            result = transfer(msgs);
+            if (UNLIKELY(result < 0))
+            {
+                /* something bad happened, reset the bus and bail */
+                csDeassert();
+                bus_unlock();
+                lock_.unlock();
```
Comment (balazsracz) [inline src/freertos_drivers/common/SPI.hxx]: nit: whenever you see "foo.unlock(); return;" sequence of commands, consider whether a RAII object would have made the code simpler, e.g. `OSMutexLock l(&lock);` at the beginning, then use just `return;` and let the destructor take care of unlocking.
Similar for portENTER_CRITICAL and portEXIT_CRITICAL pairs, using an AtomicHolder is a good alternative.
Same thing applies for break; when getting out of loops.

File: src/freertos_drivers/common/MCP2515Can.hxx [hunk 8]
```diff
@@ -330,137 +345,184 @@ private:
...
+                can_frame->can_rtr = srr_;
+                can_frame->can_id = ((uint32_t)sid_  << 0) +
+                                    ((uint32_t)sidh_ << 3);
             }
             can_frame->can_err = 0;
-            memcpy(can_frame->data, data, 8);
-            can_frame->can_dlc = dlc;
+            can_frame->can_dlc = dlc_;
+            static_assert(offsetof(Buffer, data_) == 8, "data_ misaligned");
+            *((uint64_t*)can_frame->data) = *((uint64_t*)data_);
```
Comment (balazsracz) [inline src/freertos_drivers/common/MCP2515Can.hxx]: I'm getting the following compiler error here:
```
/home/bracz/train/openlcb/tcs_booster/deps/tcs_throttle/deps/openmrn/src/freertos_drivers/common/MCP2515Can.hxx: In member function 'void MCP2515Can::Buffer::build_struct_can_frame(can_frame*)':
/home/bracz/train/openlcb/tcs_booster/deps/tcs_throttle/deps/openmrn/src/freertos_drivers/common/MCP2515Can.hxx:412:41: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
             *((uint64_t*)can_frame->data) = *((uint64_t*)data_);
```

File: src/freertos_drivers/ti/CC32xxSPI.cxx [hunk 9]
```diff
@@ -31,21 +31,27 @@
...
+#include "inc/hw_udma.h"
+#include "driverlib/gpio.h"
 #include "driverlib/rom.h"
 #include "driverlib/rom_map.h"
 #include "driverlib/spi.h"
 #include "driverlib/interrupt.h"
 #include "driverlib/prcm.h"
 
-#include "CC32xxSPI.hxx"
+#include <ti/drivers/dma/UDMACC32XX.h>
```
Comment (balazsracz) [inline src/freertos_drivers/ti/CC32xxSPI.cxx]: I'm getting the following compile error here:
```
/home/bracz/train/openlcb/tcs_booster/deps/tcs_throttle/deps/openmrn/src/freertos_drivers/ti/CC32xxSPI.cxx:51:39: fatal error: ti/drivers/dma/UDMACC32XX.h: No such file or directory
 #include <ti/drivers/dma/UDMACC32XX.h>
```
This is probably with the CC3200 SDK.

File: src/freertos_drivers/common/MCP2515Can.hxx (hunk 8 above)
Comment (balazsracz) [reply src/freertos_drivers/common/MCP2515Can.hxx]: fyi i used armgcc 2017q1 when i got this.

**Pull request merged**